src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -221,61 +254,44 @@ private bool IsSimpleImplicitExpression(CompletionParams request, LSPDocumentSna
         // We should remove Razor design time helpers from C#'s completion list. If the current identifier being targeted does not start with a double
         // underscore, we trim out all items starting with "__" from the completion list. If the current identifier does start with a double underscore
         // (e.g. "__ab[||]"), we only trim out common design time helpers from the completion list.
-        private SumType<CompletionItem[], CompletionList> RemoveDesignTimeItems(
+        private void RemoveDesignTimeItems(
             LSPDocumentSnapshot documentSnapshot,
             TextExtent? wordExtent,
-            SumType<CompletionItem[], CompletionList> completionResult)
+            CompletionList completionList)
         {
-            var result = completionResult.Match<SumType<CompletionItem[], CompletionList>>(
-                items =>
-                {
-                    items = RemoveDesignTimeItems(documentSnapshot, wordExtent, items);
-                    return items;
-                },
-                list =>
-                {
-                    list.Items = RemoveDesignTimeItems(documentSnapshot, wordExtent, list.Items);
-                    return list;
-                });
-
-            return result;
+            var filteredItems = completionList.Items.Except(DesignTimeHelpersCompletionItems, CompletionItemComparer.Instance).ToArray();
 
-            static CompletionItem[] RemoveDesignTimeItems(LSPDocumentSnapshot documentSnapshot, TextExtent? wordExtent, CompletionItem[] items)
+            // If the current identifier starts with "__", only trim out common design time helpers from the list.
+            // In all other cases, trim out both common design time helpers and all completion items starting with "__".
+            if (RemoveAllDesignTimeItems(documentSnapshot, wordExtent))
             {
-                var filteredItems = items.Except(DesignTimeHelpersCompletionItems, CompletionItemComparer.Instance).ToArray();
+                filteredItems = filteredItems.Where(item => item.Label != null && !item.Label.StartsWith("__", StringComparison.Ordinal)).ToArray();
+            }
+
+            completionList.Items = filteredItems;
 
-                // If the current identifier starts with "__", only trim out common design time helpers from the list.
-                // In all other cases, trim out both common design time helpers and all completion items starting with "__".
-                if (RemoveAllDesignTimeItems(documentSnapshot, wordExtent))
+            static bool RemoveAllDesignTimeItems(LSPDocumentSnapshot documentSnapshot, TextExtent? wordExtent)
```

Comment:
```suggestion
            static bool ShouldRemoveAllDesignTimeItems(LSPDocumentSnapshot documentSnapshot, TextExtent? wordExtent)
```

Current name seems like it will mutate the object until you look at the return type.